content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+
+### Initializer Safety
+
+Csmith initializes variables right after declaration and bans `goto` statement to ensure the execution is in order.
+
+**Question:** What kind of cases can be omitted due to the current design of program generation and safety constraints? 
+
+### Global Safety
+
+Because Csmith generates the code incrementally, newly generated code can threaten the old code.
+
+```c++
+incrementallyGeneratedUnsafeProgram(){}
+    int *p = &i;
+    while (...) {
+        *p = 3; 
+        p = 0; // unsafe because of the back-edge
+    }
+}
+```
+
+Csmith performs checking at each newly generated line except for loops. Loops are checked at the end when the back-edge is logically created. If unsafe lines appear, Csmith deletes line by line until the safety requirements are satisfied.
+
+## Design Tradeoff
+
+### Allow Implementation-defined Behavior
+
+Implementation-defined behavior is equivalent to [unspecified behavior](<https://en.wikipedia.org/wiki/Unspecified_behavior>), which may vary on different implementations of a programming language. Csmith designers believe it is unrealistic to "retain a single interpretation across all possible choices of implementation-defined behaviors". They allow Csmith to give different outputs when compilers has implementation-defined behavior at 
+
+1. the widths and representations of integers
+2. behavior when casting to a signed integer type when the value cannot be represented in an object of the target type
+3. results of bitwise operations on signed integers. 
+
+There are roughly three kinds of compiler targets:
+
+| Targets               | `int` bit length | `long` bit length |
+| --------------------- | ---------------- | ----------------- |
+| x86-64                | 32               | 64                |
+| x86, ARM, and PowerPC | 32               | 32                |
+| MSP430 and AVR        | 16               | 16                |
+
+Csmith performs testing within but not cross the equivalent classes.
+
+### No Ground Truth
+
+Csmith does not have a ground truth since it is unrealistic to have a human check each program. Instead, it takes the majority vote. In practice, two uncorrelated compilers have not output the same incorrect output, according to the authors. The explanation to this is the substantial diversity among intermediate languages (IRs).
+
+### No Guarantee of Termination
+
+A Csmith-generated program can be of any length. In practice, a time-out function is called to terminate program that takes too long to finish.
+
+### Target Middle-End bugs
```

Comment:
Use title case for all section titles. https://capra.cs.cornell.edu/styleguide/#sectitle